File: src/titan/ast/grammar/syntax/ReducingSymbol.java
package titan.ast.grammar.syntax;

import java.util.Objects;
import titan.ast.runtime.AstRuntimeException;
import titan.ast.target.AutomataTmpAst;

/**
 * 分析栈存放的文法符号.
 *
 * @author tian wei jun
 */
public class ReducingSymbol implements Cloneable, Comparable<ReducingSymbol> {
  // ast
  public AutomataTmpAst astOfCurrentDfaState = null;
  // 状态
  public SyntaxDfaState currentDfaState = null;
  // token流中的位置
  public int endIndexOfToken = -1;

  public ReducingSymbol() {}

  public ReducingSymbol clone() {
    ReducingSymbol reducingSymbol = null;
    try {
      reducingSymbol = (ReducingSymbol) super.clone();
    } catch (CloneNotSupportedException e) {
      throw new AstRuntimeException(e);
    }
    reducingSymbol.currentDfaState = this.currentDfaState;
    reducingSymbol.endIndexOfToken = this.endIndexOfToken;
    reducingSymbol.astOfCurrentDfaState = this.astOfCurrentDfaState.diyClone();
    return reducingSymbol;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    ReducingSymbol that = (ReducingSymbol) o;
    return endIndexOfToken == that.endIndexOfToken
        && astOfCurrentDfaState.equals(that.astOfCurrentDfaState)
        && currentDfaState.equals(that.currentDfaState);
  }

  @Override
  public int hashCode() {
    return Objects.hash(astOfCurrentDfaState, currentDfaState, endIndexOfToken);
  }

  @Override
  public String toString() {
    return null == astOfCurrentDfaState ? "" : astOfCurrentDfaState.toString();
  }

  @Override
  public int compareTo(ReducingSymbol that) {
    int compare = astOfCurrentDfaState.grammar.compareTo(that.astOfCurrentDfaState.grammar);
    if (0 != compare) {
      return compare;
    }
    return currentDfaState.compareTo(that.currentDfaState);
  }
}

File: runtime/rust/titan-ast-runtime/titan-ast-runtime-lib/src/reducing_symbol.rs
use std::cmp::Ordering;

use crate::automata_tmp_ast::AutomataTmpAst;

#[derive(Clone)]
pub(crate) struct ReducingSymbol {
    pub(crate) ast_of_current_dfa_state: AutomataTmpAst,
    pub(crate) current_dfa_state: usize,
    pub(crate) end_index_of_token: i32,
}

impl ReducingSymbol {
    pub(crate) fn equals(&self, other: &Self) -> bool {
        self.end_index_of_token == other.end_index_of_token
            && self.current_dfa_state == other.current_dfa_state
            && self.ast_of_current_dfa_state.get_grammar()
                == other.ast_of_current_dfa_state.get_grammar()
    }

    pub(crate) fn compare(&self, other: &Self) -> Ordering {
        if self.end_index_of_token < other.end_index_of_token {
            return Ordering::Less;
        }
        if self.end_index_of_token > other.end_index_of_token {
            return Ordering::Greater;
        }

        if self.current_dfa_state < other.current_dfa_state {
            return Ordering::Less;
        }
        if self.current_dfa_state > other.current_dfa_state {
            return Ordering::Greater;
        }

        if self.ast_of_current_dfa_state.get_grammar()
            < other.ast_of_current_dfa_state.get_grammar()
        {
            return Ordering::Less;
        }
        if self.ast_of_current_dfa_state.get_grammar()
            > other.ast_of_current_dfa_state.get_grammar()
        {
            return Ordering::Greater;
        }

        return Ordering::Equal;
    }
}

impl Default for ReducingSymbol {
    fn default() -> Self {
        Self {
            ast_of_current_dfa_state: Default::default(),
            current_dfa_state: 0,
            end_index_of_token: -1,
        }
    }
}
